config/config.go

```diff
@@ -104,6 +104,71 @@ func NewViper() *viper.Viper {
 	return viper.NewWithOptions(viper.KeyDelimiter(ViperDelimiter))
 }
 
+// fromStringMap creates a Viper object from a map[string]interface{}
+func fromStringMap(cfg map[string]interface{}) *viper.Viper {
+	viperCfg := NewViper()
+	viperCfg.MergeConfigMap(cfg)
+	return viperCfg
+}
+
+// deepSearch scans deep maps, following the key indexes listed in the
+// sequence "path".
+// The last value is expected to be another map, and is returned.
+//
+// In case intermediate keys do not exist, or map to a non-map value,
+// a new map is created and inserted, and the search continues from there:
+// the initial map "m" may be modified!
+// This function comes from Viper code https://github.com/spf13/viper/blob/5253694/util.go#L201-L230
+// It is used here because of https://github.com/spf13/viper/issues/819
+func deepSearch(m map[string]interface{}, path []string) map[string]interface{} {
+	for _, k := range path {
+		m2, ok := m[k]
+		if !ok {
+			// intermediate key does not exist
+			// => create it and continue from there
+			m3 := make(map[string]interface{})
+			m[k] = m3
+			m = m3
+			continue
+		}
+		m3, ok := m2.(map[string]interface{})
+		if !ok {
+			// intermediate key is a value
+			// => replace with a new map
+			m3 = make(map[string]interface{})
+			m[k] = m3
+		}
+		// continue search from here
+		m = m3
+	}
+	return m
+}
+
+// toStringMap creates a map[string]interface{} from a Viper object
+// It is equivalent to v.AllSettings() but it maps nil values
+// It is used here because of https://github.com/spf13/viper/issues/819
+func toStringMap(v *viper.Viper) map[string]interface{} {
+	m := map[string]interface{}{}
+	// start from the list of keys, and construct the map one value at a time
+	for _, k := range v.AllKeys() {
+		value := v.Get(k)
+		path := strings.Split(k, ViperDelimiter)
+		lastKey := strings.ToLower(path[len(path)-1])
+		deepestMap := deepSearch(m, path[0:len(path)-1])
+		// set innermost value
+		deepestMap[lastKey] = value
+	}
+	return m
+}
+
+// ToCustomUnmarshaler creates a custom unmarshaler from a Viper unmarshaled.
+// It should not be used by new code.
+func ToCustomUnmarshaler(viperUnmarshaler func(*viper.Viper, interface{}) error) component.CustomUnmarshaler {
```

Comment:
I wrote this wrapper for convenience to fix existing tests but I don't think it should be used by new components.

Reply:
Maybe not in this PR, but I would like to remove this before we declare this module stable. Maybe we can open an issue and decide how to get rid of this.

Reply:
Created #2735 to track this